### ISSUE TITLE
perf: profile LoRA training — does frozen-base eliminate the grad_clip bottleneck?

### ISSUE THREAD

**Issue by teilomillet** [teilomillet/textpolicy#42]
## Hypothesis

With LoRA, only adapter parameters (~1-5M) are trainable while the base model (~596M) is frozen. If `clip_grad_norm` only traverses trainable parameters (not frozen ones), the 62.8% grad_clip overhead may disappear entirely.

If it still traverses ALL parameters (including frozen), this is a bug/waste that should be fixed.

## Why this matters

Most real training uses LoRA/QLoRA. If the grad_clip bottleneck disappears with LoRA, then #39 (optimize grad_clip) is lower priority. If it persists, #39 is critical.

### L1: Does clip_grad_norm traverse frozen params?
```python
@pytest.mark.requires_model
def test_grad_clip_skips_frozen_params():
    """clip_grad_norm should only process trainable (non-frozen) parameters."""
    model, tokenizer = load_model('Qwen/Qwen3-0.6B')
    lora_model, _ = create_lora_setup(model, {'rank': 8, 'layers': 4})
    
    # Count trainable vs frozen
    import mlx.utils
    trainable = sum(x.size for _, x in mlx.utils.tree_flatten(lora_model.trainable_parameters()))
    total = sum(x.size for _, x in mlx.utils.tree_flatten(lora_model.parameters()))
    print(f'Trainable: {trainable/1e6:.1f}M / {total/1e6:.1f}M ({trainable/total*100:.2f}%)')
    
    # Compute gradients — should only have entries for trainable params
    # Check: does grad dict contain frozen param keys?
```

**Expected**: Trainable params ~1-5M. Gradients should only exist for trainable params. If `clip_grad_norm` receives only trainable grads, it traverses ~1M params instead of 596M.

### L2: Measure LoRA training step profile
```bash
uv run python -c "
import time, mlx.core as mx
from textpolicy.generation.mlx_generation import load_model
from textpolicy.generation.lora import create_lora_setup
from textpolicy.training.trainer import Trainer
from textpolicy.algorithms import grpo
import mlx.optimizers as optim

model, tokenizer = load_model('Qwen/Qwen3-0.6B')
lora_model, _ = create_lora_setup(model, {'rank': 8, 'layers': 4})

# Build batch, train, measure with profile=True
# Compare phase breakdown to full-model training
"
```

**Expected**: If grads are LoRA-only, grad_clip should drop from 412ms to <10ms (1M vs 596M params). Training step should be ~250ms total instead of 604ms.

### L3: LoRA vs full-model training step comparison
```python
@pytest.mark.requires_model
def test_lora_training_faster_than_full():
    """LoRA training step should be significantly faster (fewer trainable params)."""
    # Full model training step: ~604ms
    # LoRA training step: expected ~200-300ms
    # Assert lora_ms < full_ms * 0.7  (at least 30% faster)
```

### L4: LoRA grad_clip overhead
```python
@pytest.mark.requires_model
def test_lora_grad_clip_overhead():
    """Grad clip overhead with LoRA should be negligible."""
    # LoRA with clip:    X ms
    # LoRA without clip: Y ms
    # Assert (X - Y) / Y < 0.10  (less than 10% overhead)
```

## Acceptance criteria

- [ ] L1 confirms whether `nn.value_and_grad` with LoRA produces LoRA-only gradients
- [ ] L2 provides real LoRA training profile data
- [ ] L3 quantifies LoRA vs full-model training speed
- [ ] L4 quantifies LoRA grad_clip overhead specifically
- [ ] Results documented to guide priority of #39
